utils/plugins/utils.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/jfrog/jfrog-cli-core/v2/utils/lock"
+	cliLog "github.com/jfrog/jfrog-cli-core/v2/utils/log"
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+	"github.com/jfrog/jfrog-client-go/utils/io/fileutils"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+)
+
+// Internal golang locking for the same process.
+var mutex sync.Mutex
+
+func init() {
+	cliLog.SetDefaultLogger()
+}
+
+type PluginsV1 struct {
+	Version int `json:"version,omitempty"`
+}
+
+func CheckPluginsVersionAndConvertIfNeeded() error {
+	// Locking mechanism
+	mutex.Lock()
+	defer mutex.Unlock()
+	lockDirPath, err := coreutils.GetJfrogPluginsLockDir()
+	if err != nil {
+		return err
+	}
+	lockFile, err := lock.CreateLock(lockDirPath)
+	defer lockFile.Unlock()
+	if err != nil {
+		return err
+	}
+	// Check if 'plugins' directory exists in .jfrog
+	jfrogHomeDir, err := coreutils.GetJfrogHomeDir()
+	if err != nil {
+		return err
+	}
+	exists, err := fileutils.IsDirExists(filepath.Join(jfrogHomeDir, coreutils.JfrogPluginsDirName), false)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return nil
+	}
+
+	plugins, err := readPluginsYaml()
+	if err != nil {
+		return err
+	}
+	if plugins.Version != coreutils.GetPluginsYamlVersion() {
+		return errorutils.CheckError(errors.New(fmt.Sprintf("Expected plugins version in 'plugins.yaml is %d  but the actual value is %d", coreutils.GetPluginsYamlVersion(), plugins.Version)))
+	}
+	return nil
+}
+
+func readPluginsYaml() (*PluginsV1, error) {
+	plugins := new(PluginsV1)
+	content, err := getPluginsYamlFile()
+	if err != nil {
+		return nil, err
+	}
+	if len(content) == 0 {
+		// No plugins.yaml file was found, that means that we are in v0.
+		// Convert plugins layout to the latest version.
+		return convertPluginsV0ToV1()
+	}
+
+	err = json.Unmarshal(content, &plugins)
+	if err != nil {
+		return nil, errorutils.CheckError(err)
+	}
+	return plugins, err
+}
+
+func getPluginsYamlFile() (content []byte, err error) {
+	pluginsFilePath, err := getPluginsFilePath()
+	if err != nil {
+		return
+	}
+	exists, err := fileutils.IsFileExists(pluginsFilePath, false)
+	if err != nil {
+		return
+	}
+	if exists {
+		content, err = fileutils.ReadFile(pluginsFilePath)
+	}
+	return
+}
+
+// V0: in plugins directory there was no 'plugins.yaml' file. all executable files were in the same directory.
+// V1: create 'plugins.yaml' file inside 'plugins' directory. change the file's hierarchy inside 'plugins' directory.
+func convertPluginsV0ToV1() (*PluginsV1, error) {
+	err := migrateFileSystemLayoutV0ToV1()
+	if err != nil {
+		return nil, err
+	}
+	return CreatePluginsYamlFile()
+}
+
+// Change the file's hierarchy inside 'plugins' directory to:
+//	plugins (dir)
+//		plugin-name (dir)
+//			bin (dir)
+//				plugin-executable (file)
+//			resources:(optional dir)
+//				... (directories/files)
+func migrateFileSystemLayoutV0ToV1() error {
```

Comment:
migrateFileSystemLayoutV0ToV1 --> convertFileSystemLayoutV0ToV1
(since we're already using the tern "convert" in the code, we better be consistent and not use "migrate".